upf/upf_pfcp_api.c

```diff
@@ -687,6 +731,66 @@ validate_ue_ip (upf_session_t * sx, upf_nwi_t * nwi,
     response->failed_rule_id.id = pdr->pdr_id;				\
   } while (0)
 
+upf_nat_addr_t *
+get_nat_addr (upf_nat_pool_t * np)
+{
+  upf_nat_addr_t *this_addr = NULL, *addr = NULL;
+  u32 least_locked_ports = ~0;
+  vec_foreach (this_addr, np->addresses)
```

Comment:
this could be very expensive when using large pools

Reply:
Yes, it's true. For phase 2 we have I assumed to have simple pool management algorithm (but still it handles more than 1K establishment requests per sec on a /18 networks)